store/helper/helper.go

```diff
@@ -399,36 +412,208 @@ type RegionsInfo struct {
+func isBeforeKeyRange(x withKeyRange, startKey, endKey string) bool {
+	return x.getEndKey() != "" && x.getEndKey() <= startKey
+}
+
+// IsBehind returns true is x is behind [startKey, endKey)
+func isBehindKeyRange(x withKeyRange, startKey, endKey string) bool {
+	return endKey != "" && x.getStartKey() >= endKey
+}
+
+func (r *RegionInfo) getStartKey() string { return r.StartKey }
+func (r *RegionInfo) getEndKey() string   { return r.EndKey }
+
+// for sorting
+type byRegionStartKey []*RegionInfo
+
+func (xs byRegionStartKey) Len() int      { return len(xs) }
+func (xs byRegionStartKey) Swap(i, j int) { xs[i], xs[j] = xs[j], xs[i] }
+func (xs byRegionStartKey) Less(i, j int) bool {
+	return xs[i].getStartKey() < xs[j].getStartKey()
+}
+
+// tableInfoWithKeyRange stores table or index informations with its key range.
+type tableInfoWithKeyRange struct {
+	*TableInfo
+	StartKey string
+	EndKey   string
+}
+
+func (t tableInfoWithKeyRange) getStartKey() string { return t.StartKey }
+func (t tableInfoWithKeyRange) getEndKey() string   { return t.EndKey }
+
+// for sorting
+type byTableStartKey []tableInfoWithKeyRange
+
+func (xs byTableStartKey) Len() int      { return len(xs) }
+func (xs byTableStartKey) Swap(i, j int) { xs[i], xs[j] = xs[j], xs[i] }
+func (xs byTableStartKey) Less(i, j int) bool {
+	return xs[i].getStartKey() < xs[j].getStartKey()
+}
+
+func newTableWithKeyRange(db *model.DBInfo, table *model.TableInfo) tableInfoWithKeyRange {
+	sk, ek := tablecodec.GetTableHandleKeyRange(table.ID)
+	startKey := bytesKeyToHex(codec.EncodeBytes(nil, sk))
+	endKey := bytesKeyToHex(codec.EncodeBytes(nil, ek))
+	return tableInfoWithKeyRange{
+		&TableInfo{
+			DB:      db,
+			Table:   table,
+			IsIndex: false,
+			Index:   nil,
+		},
+		startKey,
+		endKey,
+	}
+}
+
+func newIndexWithKeyRange(db *model.DBInfo, table *model.TableInfo, index *model.IndexInfo) tableInfoWithKeyRange {
+	sk, ek := tablecodec.GetTableIndexKeyRange(table.ID, index.ID)
+	startKey := bytesKeyToHex(codec.EncodeBytes(nil, sk))
+	endKey := bytesKeyToHex(codec.EncodeBytes(nil, ek))
+	return tableInfoWithKeyRange{
+		&TableInfo{
+			DB:      db,
+			Table:   table,
+			IsIndex: true,
+			Index:   index,
+		},
+		startKey,
+		endKey,
+	}
+}
+
+// GetRegionsTableInfo returns a map maps region id to its tables or indices.
+// Assuming tables or indices key ranges never intersect.
+// Regions key ranges can intersect.
+func (h *Helper) GetRegionsTableInfo(regionsInfo *RegionsInfo, schemas []*model.DBInfo) map[int64][]TableInfo {
+	tableInfos := make(map[int64][]TableInfo)
+
+	regions := []*RegionInfo{}
+	for i := 0; i < len(regionsInfo.Regions); i++ {
+		tableInfos[regionsInfo.Regions[i].ID] = []TableInfo{}
+		regions = append(regions, &regionsInfo.Regions[i])
+	}
+
+	tables := []tableInfoWithKeyRange{}
+	for _, db := range schemas {
+		for _, table := range db.Tables {
+			tables = append(tables, newTableWithKeyRange(db, table))
```

Comment:
Can we reuse the function `FindTableIndexOfRegion`?

Reply:
That function may cause performance issues. It uses `O(M)` time to find table infos for one region, where `M` is the number of tables and indices. So, it takes `O(MN)` time to find table infos for `N` regions. My implementation only need `O(M log(M) + N log(N) + kN)` time, where `k` is the average number of tables and indices in one region(which is usually very small). For hot regions (`N` is not very large), maybe `O(MN)` is not a problem, but this PR is working on all regions.
What's more, `FindTableIndexOfRegion`  assumes there is only one table or index in one region, which is not true (I'm not sure is this a bug).

Reply:
`FindTableIndexOfRegion` is copied from the old codes in package `server`.

Reply:
> `FindTableIndexOfRegion` is copied from the old codes in package `server`. If the assumption is break, you can change the implementation of it.

It seems that will change a lot of old codes. I think that should be another PR.